### PR TITLE
Adding support for JDK17 and removing JDK8

### PR DESCRIPTION
**Pull request by martin-gaievski**
### Description
Making necessary adjustments to make build compatible with JDK17:
- add `add-exports` arguments due to `java.lang.IllegalAccessError` when `spotlessJavaCheck` uses classes from `com.sun.*`
- remove bad javadoc formatting as this generates warnings and build is running with `-Werror` option

Changing runtime java version from 8 to 11.
 
### Issues Resolved
[Remove support for Java 8, add support for Java 17](https://github.com/opensearch-project/opensearch-plugins/issues/110)
 
### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dreamer-89** [comment]
Gradle check is failing `testWithKeywordLongAndMissingBucket` test which seems flaky. Its worth retyring Gradle check again.
```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.testWithKeywordLongAndMissingBucket" -Dtests.seed=4E1342EC6B964D83 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=zh-Hans-CN -Dtests.timezone=Europe/Isle_of_Man -Druntime.java=17

org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests > testWithKeywordLongAndMissingBucket FAILED
    org.junit.ComparisonFailure: expected:<{keyword=null, long=[null]}> but was:<{keyword=null, long=[100]}>
        at __randomizedtesting.SeedInfo.seed([4E1342EC6B964D83:AF2CA28B0C4823CC]:0)
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.lambda$testWithKeywordLongAndMissingBucket$85(CompositeAggregatorTests.java:1057)
        at org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.executeTestCase(CompositeAggregatorTests.java:2712)
        at org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.testSearchCase(CompositeAggregatorTests.java:2660)
        at org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.testWithKeywordLongAndMissingBucket(CompositeAggregatorTests.java:1045)
```

File: gradle.properties [hunk 1]
```diff
@@ -11,7 +11,12 @@
 
 org.gradle.warning.mode=none
 org.gradle.parallel=true
-org.gradle.jvmargs=-Xmx3g -XX:+HeapDumpOnOutOfMemoryError -Xss2m
+org.gradle.jvmargs=-Xmx3g -XX:+HeapDumpOnOutOfMemoryError -Xss2m \
+    --add-exports jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED \
```
Comment (reta) [inline gradle.properties:15]: I am wondering why these exports are needed? AFAIK OpenSearch does not use Java Compiler API, do you have an example at hand? Thank you.

**Comment by martin-gaievski** [comment]
> Gradle check is failing `testWithKeywordLongAndMissingBucket` test which seems flaky. Its worth retyring Gradle check again.
> 
> ```
> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.testWithKeywordLongAndMissingBucket" -Dtests.seed=4E1342EC6B964D83 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=zh-Hans-CN -Dtests.timezone=Europe/Isle_of_Man -Druntime.java=17
> 
> org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests > testWithKeywordLongAndMissingBucket FAILED
>     org.junit.ComparisonFailure: expected:<{keyword=null, long=[null]}> but was:<{keyword=null, long=[100]}>
>         at __randomizedtesting.SeedInfo.seed([4E1342EC6B964D83:AF2CA28B0C4823CC]:0)
>         at org.junit.Assert.assertEquals(Assert.java:117)
>         at org.junit.Assert.assertEquals(Assert.java:146)
>         at org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.lambda$testWithKeywordLongAndMissingBucket$85(CompositeAggregatorTests.java:1057)
>         at org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.executeTestCase(CompositeAggregatorTests.java:2712)
>         at org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.testSearchCase(CompositeAggregatorTests.java:2660)
>         at org.opensearch.search.aggregations.bucket.composite.CompositeAggregatorTests.testWithKeywordLongAndMissingBucket(CompositeAggregatorTests.java:1045)
> ```

hey Suraj, 
How can I restart the check (Gradle check that you've mentioned)? Checks that are [listed in PR's check tab](https://github.com/opensearch-project/OpenSearch/pull/2025/checks) are all green.
^ @dreamer-89

**Comment by owaiskazi19** [comment]
start gradle check

File: gradle.properties (hunk 1 above)
Comment (reta) [reply gradle.properties:15]: Ah ... thank you @martin-gaievski , `spotless` again ...

**Comment by reta** [comment]
> x   Gradle Check failure [988b457](https://github.com/opensearch-project/OpenSearch/commit/988b457d0959dd0ffbb15a1347d5068e519bc00f) [Log 2150](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_2150.log)
> 
> [Reports 2150](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_2150_reports.zip)

Hehe ... @owaiskazi19 seems like without our https://github.com/opensearch-project/OpenSearch/pull/2024 PR merged, not much hope for builds to succeed :( 

```
* What went wrong:
Execution failed for task ':distribution:docker:buildDockerImage'.
> A failure occurred while executing org.opensearch.gradle.docker.DockerBuildTask$DockerBuildAction
   > Process 'command 'docker'' finished with non-zero exit value 1
```

**Comment by owaiskazi19** [comment]
> 

Yeah. Looks like all the build failed for the same reason.

**Comment by owaiskazi19** [comment]
> > x   Gradle Check failure [988b457](https://github.com/opensearch-project/OpenSearch/commit/988b457d0959dd0ffbb15a1347d5068e519bc00f) [Log 2150](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_2150.log)
> > [Reports 2150](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_2150_reports.zip)
> 
> Hehe ... @owaiskazi19 seems like without our #2024 PR merged, not much hope for builds to succeed :(
> 
> ```
> * What went wrong:
> Execution failed for task ':distribution:docker:buildDockerImage'.
> > A failure occurred while executing org.opensearch.gradle.docker.DockerBuildTask$DockerBuildAction
>    > Process 'command 'docker'' finished with non-zero exit value 1
> ```

Yeah. Looks like all the build failed for the same reason.

**Comment by owaiskazi19** [comment]
start gradle check

**Comment by dblock** [comment]
```
Tests with failures:
 - org.opensearch.cluster.routing.MovePrimaryFirstTests.testClusterGreenAfterPartialRelocation
 ````

https://github.com/opensearch-project/OpenSearch/issues/1957

**Comment by dblock** [comment]
This overlaps with https://github.com/opensearch-project/OpenSearch/pull/1977. 

@martin-gaievski want to merge these two and update docs?

**Comment by martin-gaievski** [comment]
> This overlaps with #1977.
> 
> @martin-gaievski want to merge these two and update docs?

sure, added changes from #1977 so now this PR consolidates all changes
^ @dblock

**Comment by andrross** [comment]
The [developer guide](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md?plain=1#L6) needs to be updated as well.

**Comment by martin-gaievski** [comment]
> 

Ack, updated in next commit

File: DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -65,13 +65,13 @@ OpenSearch builds using Java 11 at a minimum. This means you must have a JDK 11
 
 Download Java 11 from [here](https://adoptium.net/releases.html?variant=openjdk11).
 
-#### JDK 8 and 14
+#### JDK 14
 
-To run the full suite of tests, download and install [JDK 8](https://adoptium.net/releases.html?variant=openjdk8) and [JDK 14](https://jdk.java.net/archive/) and set `JAVA8_HOME`, `JAVA11_HOME`, and `JAVA14_HOME`. They are required by the [backwards compatibility test](./TESTING.md#testing-backwards-compatibility).
+To run the full suite of tests, download and install [JDK 14](https://jdk.java.net/archive/) and set `JAVA11_HOME`, and `JAVA14_HOME`. They are required by the [backwards compatibility test](./TESTING.md#testing-backwards-compatibility).
```
Comment (reta) [inline DEVELOPER_GUIDE.md:70]: @martin-gaievski this is interesting, we are not going to backport this change to 1.x, right? In this case, I am not sure we have to drop the JDK-8 requirements for bwc tests, did you have a chance to check it?

Comment (martin-gaievski) [reply DEVELOPER_GUIDE.md:70]: @reta As to my knowledge this is not going to be back-ported to 1.x. It's meant to be for 2.0 only as per original requirements from [110](https://github.com/opensearch-project/opensearch-plugins/issues/110). 
I've run bwcTest locally for this branch with only JAVA11_HOME and JAVA14_HOME set and it passed without errors.

Comment (reta) [reply DEVELOPER_GUIDE.md:70]: All set, thank you

**Review by dblock — APPROVED** [review]
Nice work @martin-gaievski, thanks!

**Pull request merged**